src/compiler/tsbuildPublic.ts

```diff
@@ -1609,8 +1609,14 @@ namespace ts {
 
             if (buildInfo.program) {
                 // If there are pending changes that are not emitted, project is out of date
+                // If noEmit, then explicitly check if there are semantic diagnostics
+                // affectedFilesPendingEmit is present in noEmit irrespective of errors to handle files to be emitted when noEmit is false,

```

Comment:
This helps me understand why `affectedFilesPendingEmit` behaves differently under `noEmit`, but I don’t understand why that makes checking for the presence of semantic errors the correct fallback. Why does the presence of semantic errors definitely mean the buildinfo is out of date? Why isn’t the same true of syntactic errors?

Reply:
If program has syntactic errors the `changeFileSet` because semantic diagnostics are not even processed in that case and state remains same. (https://github.com/microsoft/TypeScript/pull/50974/files#diff-2461280d51916bf4023650d8950c0af3bf16011e3f0c8a290c7c9f70056147a0R77) I will update the comment